misk/src/main/kotlin/misk/client/TypedHttpClientModule.kt

```diff
@@ -101,3 +103,96 @@ class TypedHttpClientModule<T : Any>(
     }
   }
 }
+
+/**
+ * Factory for creating typed clients that call other members of a cluster.
+ */
+interface TypedPeerClientFactory<T> {
+  fun client(peer: Cluster.Member): T
+}
+
+/**
+ * Creates a retrofit-backed typed client factory given an API interface and an HTTP configuration.
+ *
+ * The factory returned typed clients that can be used to call other members of the cluster.
+ */
+class TypedPeerHttpClientModule<T : Any>(
+  private val kclass: KClass<T>,
+  private val name: String
+) : KAbstractModule() {
+
+  override fun configure() {
+    // Initialize empty sets for our multibindings.
+    newMultibinder<ClientNetworkInterceptor.Factory>()
+    newMultibinder<ClientApplicationInterceptor.Factory>()
+
+    @Suppress("UNCHECKED_CAST")
+    val key = Key.get(
+        Types.newParameterizedType(TypedPeerClientFactory::class.java,
+            kclass.java)) as Key<TypedPeerClientFactory<T>>
+
+    bind(key).toProvider(PeerTypedClientProvider(kclass, name))
+  }
+
+  companion object {
+    inline fun <reified T : Any> create(name: String): TypedPeerHttpClientModule<T> {
+      return TypedPeerHttpClientModule(T::class, name)
+    }
+  }
+
+  private class PeerTypedClientProvider<T : Any>(
```

Comment:
Adapted (err, cargo culted) from `TypedHttpClientModule`

Reply:
yea this feels copied. can we do better and share? these configs/bindings are especially tricky to get right. e.g that stupid provider cycle detector note below